util/py/scripts/bazel_aspect_tool_audit_sec_mmio_calls.py

```diff
@@ -0,0 +1,112 @@
+from util.py.packages.lib.register_usage_report import RegisterUsageReportGroup
+
+# Define analyzers for each of the sec_mmio_* functions we care about. We give
+# each analyzer a list of token patterns where None is a stand-in for a register
+# offset.
+#
+# If the list of token patterns is incomplete, we will fail the Bazel build by
+# raising an uncaught exception.
+#
+# NOTE: This list is likely to become out of date as the codebase evolves.
+SEC_MMIO_FUNCTION_ANALYZERS = [
+    CallSiteAnalyzer(
+        function_name="sec_mmio_read32",
+        arg_index=0,
+        reg_token_patterns=[
+            TokPat(['0']),
+            TokPat(['4']),
+            TokPat(['8']),
+            TokPat(['cfg_addr']),
+            TokPat('kBase + reg_offset + sizeof ( uint32_t )'.split()),
+            TokPat('kBase + reg_offset + i * sizeof ( uint32_t )'.split()),
+            TokPat(['kBase', '+', None, '+', 'address']),
+            TokPat([
+                'kBase', '+', None, '+', 'i', '*', 'sizeof', '(', 'uint32_t',
+                ')'
+            ]),
+            TokPat(['kBase', '+', None]),
+            TokPat(['info_page', '->', 'cfg_addr']),
+        ]),
+    CallSiteAnalyzer(function_name="sec_mmio_write32",
+                     arg_index=0,
+                     reg_token_patterns=[
+                         TokPat(['0']),
+                         TokPat(['4']),
+                         TokPat(['8']),
+                         TokPat(['cfg_addr']),
+                         TokPat([
+                             'kBase', '+', None, '+', 'i', '*', 'sizeof', '(',
+                             'uint32_t', ')'
+                         ]),
+                         TokPat(['kBase', '+', None]),
+                         TokPat(['kPwrMgrBase', '+', None]),
+                         TokPat(['regs', '.', 'cfg_addr']),
+                         TokPat(['regs', '.', 'cfg_wen_addr']),
+                         TokPat(['info_page', '->', 'cfg_addr']),
+                         TokPat(['info_page', '->', 'cfg_wen_addr']),
+                     ]),
+    CallSiteAnalyzer(function_name="sec_mmio_write32_shadowed",
+                     arg_index=0,
+                     reg_token_patterns=[
+                         TokPat(['kBase', '+', None]),
+                     ])
+]
+
+
+def main():
+    generated_file = Path(sys.argv[1])
+    source_file = Path(sys.argv[2])
+    clang_args = sys.argv[3:]
+
+    assert not generated_file.exists()
+    assert source_file.exists()
+
+    index = clang.cindex.Index.create()
+
+    try:
+        translation_unit = index.parse(source_file, args=clang_args)
+    except clang.cindex.TranslationUnitLoadError as e:
+        # Something went wrong inside of libclang, but it is not forthcoming
+        # with details. Manually run `clang` and hope that it produces a more
+        # helpful error message.
+        command = ["clang"] + clang_args + [source_file]
+        print("command:", command)
```

Comment:
nit: would logging an error instead of printing here work?

Reply:
Ah, I actually used print intentionally here. There's a couple reasons:

* This script (bazel_aspect_tool_audit_sec_mmio_calls.py) is never directly called by a human in normal use, so there's no way for the user to change the logging level.
* I wanted to print the command immediately before the output it generates. The subproces.run() call will write directly to stdout/stderr.

But at the end of the day, it's just a convenience feature for debugging this script. I'm also happy to delete the print if you feel strongly.